### PR TITLE
PMM-10705 Installing operators via OLM

### PR DESCRIPTION
**Pull request by percona-csalguero**
This is a POC to install OLM and replace hardcoded parts like PXC and
PSMDB handlers.
I added some methods to the K8sclient package that only expose methods
that already exist in the kubectl package.
I would like to replace the K8sClient and remove all method that are
specific for specific operators like PXC or PSMDB. I would have a
K8sUtil library with general methods and I would try to make it possible
to handle ANY operator with our library.
Also, kubectl shouldn't be internal, it must be possible to use it from
any other package. That's why I exposed some methods like Apply or Run
in K8sClient, because otherwsie I should write OLM specific code into
K8sClient. If it is K8s Client, it should only containt K8s generic
code, not code for specific operators/operator versions.

PMM-0

Build: SUBMODULES-0

- [ ] Links to other linked pull requests (optional).

### CODE REVIEW

**Comment by denisok** [comment]
I checked the code and it look good. So far I understand that "update" feature is missed and operators update would be automatic, not manual ?

Also we would need to discuss Namespaces. To understand standards and practices for operator deployments, and how they operate in namespaces versus "cluster wide".

**Comment by percona-csalguero** [comment]
Yes, first attempt was with automatic updates just because update is missing.

**Review by denisok — APPROVED** [review]

File: deploy/olm/pxc/percona-xtradb-cluster-operator.yaml [hunk 1]
```diff
@@ -0,0 +1,24 @@
+apiVersion: v1
```
Comment (gen1us2k) [inline deploy/olm/pxc/percona-xtradb-cluster-operator.yaml:1]: Will we drop these files and move them to the PMM once we'll have a catalog available?

Comment (percona-csalguero) [reply deploy/olm/pxc/percona-xtradb-cluster-operator.yaml:1]: Deleted.

File: service/k8sclient/k8sclient.go [hunk 2]
```diff
@@ -413,6 +421,22 @@ func (c *K8sClient) Cleanup() error {
...
+}
+
+func (c *K8sClient) Patch(ctx context.Context, patchType kubectl.PatchType, resourceType, resourceName, namespace string, res interface{}) error {
+	return c.kubeCtl.Patch(ctx, patchType, resourceType, resourceName, namespace, res)
+}
+
+func (c *K8sClient) Delete(ctx context.Context, res interface{}) error {
+	return c.kubeCtl.Delete(ctx, res)
+}
+
```
Comment (gen1us2k) [inline service/k8sclient/k8sclient.go:439]: We need to restore Cleanup logic here and call c.kubeCtl.Cleanup

**Review by gen1us2k — COMMENTED** [review]
Let's fix the pipeline and merge it

**Review by gen1us2k — APPROVED** [review]

File: deploy/olm/olm.yaml [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+      version: v1
+      kind: PackageManifest
+      name: packagemanifests
+      displayName: PackageManifest
+      description: A PackageManifest is a resource generated from existing CatalogSources and their ConfigMaps
+      deploymentName: packageserver
+      containerPort: 5443
+---
+apiVersion: operators.coreos.com/v1alpha1
+kind: CatalogSource
```
Comment (denisok) [inline deploy/olm/olm.yaml:340]: @percona-csalguero should we switch to our source ?
https://github.com/percona/dbaas-catalog/blob/percona-platform/percona-dbaas-catalog.yaml

or it is coming later?

File: service/olm/olm_operator.go [hunk 4]
```diff
@@ -0,0 +1,568 @@
...
+	var crdFile, olmFile interface{}
+
+	switch strings.ToLower(req.Version) {
+	case latestOLMVersion:
+		latestVersion, err := getLatestVersion(ctx, olmRepo)
+		if err != nil {
+			return nil, err
+		}
+		crdFile = "https://" + path.Join(baseDownloadURL, latestVersion, "crds.yaml")
+		olmFile = "https://" + path.Join(baseDownloadURL, latestVersion, "olm.yaml")
```
Comment (denisok) [inline service/olm/olm_operator.go:103]: yeah, default olm would come with default CatalogSource. We would need to switch it to the https://github.com/percona/dbaas-catalog/blob/percona-platform/percona-dbaas-catalog.yaml. Don't know if we plan it now or later?

**Comment by percona-csalguero** [comment]
I've created an new EKS cluster like this: 
```
eksctl create cluster --write-kubeconfig  --node-type=m5.xlarge --nodes=5
```
Set the env variable to use a specific `kubectl` version:
```
export KUBECTL_CMD=/home/user/kubectl-1.23
```
Then ran the tests: 
```
make test
go test  -race -p 1 -timeout=30m ./...
ok      github.com/percona-platform/dbaas-controller    0.353s
?       github.com/percona-platform/dbaas-controller/cmd/dbaas-controller       [no test files]
ok      github.com/percona-platform/dbaas-controller/service/cluster    0.087s
ok      github.com/percona-platform/dbaas-controller/service/k8sclient  777.395s
ok      github.com/percona-platform/dbaas-controller/service/k8sclient/common   0.034s
ok      github.com/percona-platform/dbaas-controller/service/k8sclient/internal/kube    17.359s
?       github.com/percona-platform/dbaas-controller/service/k8sclient/internal/kube/psmdb      [no test files]
?       github.com/percona-platform/dbaas-controller/service/k8sclient/internal/kube/pxc        [no test files]
ok      github.com/percona-platform/dbaas-controller/service/k8sclient/internal/kubectl 1.311s
?       github.com/percona-platform/dbaas-controller/service/k8sclient/internal/monitoring      [no test files]
ok      github.com/percona-platform/dbaas-controller/service/logs       0.070s
ok      github.com/percona-platform/dbaas-controller/service/olm        225.273s
?       github.com/percona-platform/dbaas-controller/utils/app  [no test files]
ok      github.com/percona-platform/dbaas-controller/utils/convertors   0.020s
?       github.com/percona-platform/dbaas-controller/utils/logger       [no test files]
ok      github.com/percona-platform/dbaas-controller/utils/servers      0.031s
?       github.com/percona-platform/dbaas-controller/utils/testutil     [no test files]
```

Maybe there is something on the cluster we use for CI?
@gen1us2k Could you check it please?

**Pull request merged**